lib/compress/zstd_preSplit.c

```diff
@@ -0,0 +1,157 @@
+    size_t n;
+    assert(s >= HASHLENGTH);
+    for (n = 0; n < limit; n++) {
+        fp->events[hash2(p++)]++;
+    }
+    fp->nbEvents += limit;
+}
+
+static void recordFingerprint(FingerPrint* fp, const void* src, size_t s)
+{
+    ZSTD_memset(fp, 0, sizeof(*fp));
+    addToFingerprint(fp, src, s);
+}
+
+static S64 abs64(S64 i) { return (i < 0) ? -i : i; }
+
+static S64 fpDistance(const FingerPrint* fp1, const FingerPrint* fp2)
+{
+    S64 distance = 0;
+    size_t n;
+    for (n = 0; n < HASHTABLESIZE; n++) {
+        distance +=
+            abs64(fp1->events[n] * fp2->nbEvents - fp2->events[n] * fp1->nbEvents);
+    }
+    return distance;
+}
+
+/* Compare newEvents with pastEvents
+ * return 1 when considered "too different"
+ */
+static int compareFingerprints(const FingerPrint* ref,
+                            const FingerPrint* newfp,
+                            int penalty)
+{
+    assert(ref->nbEvents > 0);
+    assert(newfp->nbEvents > 0);
+    {   S64 p50 = ref->nbEvents * newfp->nbEvents;
+        S64 deviation = fpDistance(ref, newfp);
+        S64 threshold = p50 * (THRESHOLD_BASE + penalty) / THRESHOLD_PENALTY_RATE;
+        return deviation >= threshold;
+    }
+}
+
+static void mergeEvents(FingerPrint* acc, const FingerPrint* newfp)
+{
+    size_t n;
+    for (n = 0; n < HASHTABLESIZE; n++) {
+        acc->events[n] += newfp->events[n];
+    }
+    acc->nbEvents += newfp->nbEvents;
+}
+
+static void flushEvents(FPStats* fpstats)
+{
+    size_t n;
+    for (n = 0; n < HASHTABLESIZE; n++) {
+        fpstats->pastEvents.events[n] = fpstats->newEvents.events[n];
+    }
+    fpstats->pastEvents.nbEvents = fpstats->newEvents.nbEvents;
+    ZSTD_memset(&fpstats->newEvents, 0, sizeof(fpstats->newEvents));
+}
+
+static void removeEvents(FingerPrint* acc, const FingerPrint* slice)
+{
+    size_t n;
+    for (n = 0; n < HASHTABLESIZE; n++) {
+        assert(acc->events[n] >= slice->events[n]);
+        acc->events[n] -= slice->events[n];
+    }
+    acc->nbEvents -= slice->nbEvents;
+}
+
+#define CHUNKSIZE (8 << 10)
+/* Note: technically, we use CHUNKSIZE, so that's 8 KB */
+size_t ZSTD_splitBlock_4k(const void* src, size_t srcSize,
+                        size_t blockSizeMax,
+                        void* workspace, size_t wkspSize)
+{
+    FPStats* const fpstats = (FPStats*)workspace;
+    const char* p = (const char*)src;
+    int penalty = THRESHOLD_PENALTY;
+    size_t pos = 0;
+    if (srcSize <= blockSizeMax) return srcSize;
+    assert(blockSizeMax == (128 << 10));
+    assert(workspace != NULL);
+    assert((size_t)workspace % ZSTD_ALIGNOF(FPStats) == 0);
+    ZSTD_STATIC_ASSERT(ZSTD_SLIPBLOCK_WORKSPACESIZE == sizeof(FPStats));
+    assert(wkspSize >= sizeof(FPStats)); (void)wkspSize;
+
+    initStats(fpstats);
+    recordFingerprint(&fpstats->pastEvents, p, CHUNKSIZE);
+    for (pos = CHUNKSIZE; pos < blockSizeMax; pos += CHUNKSIZE) {
+        assert(pos <= blockSizeMax - CHUNKSIZE);
```

Comment:
I know this holds because `blockSizeMax == (128 << 10)`, but can we fix the looping logic to guarantee this instead?

It would be cleaner to have this function support any `blockSizeMax`. Given that the checks in outer loop isn't super performance critical, I'd prefer clarity and explicit checks rather than assumptions.

Reply:
done